python/lsst/resources/_resourceHandles/_s3ResourceHandle.py

```diff
@@ -0,0 +1,260 @@
+        self._last_flush_position = self._buffer.tell() + (self._last_flush_position or 0)
+        self._buffer = BytesIO()
+
+    @property
+    def isatty(self) -> bool:
+        return False
+
+    def readable(self) -> bool:
+        return self._readable
+
+    def readline(self, size: int = -1) -> bytes:
+        raise OSError("S3 Does not support line by line reads")
+
+    def readlines(self, hint: int = -1) -> Iterable[bytes]:
+        self.seek(0)
+        return self.read().split(self._newline)
+
+    def seek(self, offset: int, whence: int = SEEK_SET) -> int:
+        if self.writable():
+            if self._last_flush_position is not None:
+                if whence == SEEK_SET:
+                    offset -= self._last_flush_position
+                    if offset < 0:
+                        raise OSError("S3 ResourceHandle can not seek prior to already flushed positions")
+                if whence == SEEK_CUR:
+                    if (self.tell() - self._last_flush_position) < 0:
+                        raise OSError("S3 ResourceHandle can not seek prior to already flushed positions")
+                if whence == SEEK_END:
+                    raise OSError("S3 ResourceHandle can not seek referencing the end of the resource")
+            self._buffer.seek(offset, whence)
+            self._position = self._buffer.tell()
+        else:
+            if whence == SEEK_SET:
+                self._position = offset
+            elif whence == SEEK_CUR:
+                self._position += offset
+            elif whence == SEEK_END:
+                offset = abs(offset)
+                self._position -= offset
+        return self._position
+
+    def seekable(self) -> bool:
+        return True
+
+    def truncate(self, size: Optional[int] = None) -> int:
+        if self.writable():
+            self._buffer.truncate(size)
+            return self._position
+        else:
+            raise OSError("ResourceHandle is not writable")
+
+    def writable(self) -> bool:
+        return self._writable
+
+    def writelines(self, lines: Iterable[bytes]) -> None:
+        if self.writable():
+            self._buffer.writelines(lines)
+            self._position = self._buffer.tell()
+        else:
+            raise OSError("ResourceHandle is not writable")
+
+    def read(self, size: int = -1) -> bytes:
+        if not self.readable():
+            raise OSError("ResourceHandle is not readable")
+        # If the object is rw, then read from the internal io buffer
+        if "+" in self._mode:
+            self._buffer.seek(self._position)
+            return self._buffer.read(size)
+        # otherwise fetch the appropriate bytes from the remote resource
+        if size > 0:
+            stop = f"{self._position + size - 1}"
+        else:
+            stop = ""
+        args = {"Range": f"bytes={self._position}-{stop}"}
+        response = self._client.get_object(Bucket=self._bucket, Key=self._key, **args)
```

Comment:
See in s3.py but there are going to have to be try blocks around all the S3 API calls or backoff handlers. (there is also the `_translate_client_error` kerfuffle in that file as well.